linalg/dual.hpp

```diff
@@ -0,0 +1,408 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+/**
+ * @file dual.hpp
+ *
+ * @brief This file contains the declaration of a dual number class
+ */
+
+#pragma once
+
+#include <cmath>
+#include "../general/backends.hpp"
+
+namespace mfem
+{
+namespace internal
+{
+
+/**
+ * @brief Dual number struct (value plus gradient)
+ * @tparam gradient_type The type of the gradient (should support addition,
+ * scalar multiplication/division, and unary negation operators)
+ */
+template <typename value_type, typename gradient_type>
+struct dual
+{
+   /// the actual numerical value
+   value_type value;
+   /// the partial derivatives of value w.r.t. some other quantity
+   gradient_type gradient;
+
+   /** @brief assignment of a double to a value of a dual. Promotes a double to
+    * a dual with a zero gradient value. */
+   auto operator=(double a) -> dual<value_type, gradient_type>&
```

Comment:
Some of the other operator definitions do require the trailing return type syntax (in C++11 anyway). I think it's preferable to have all the operator overload implementations have a consistent form, rather than switch back and forth. 

Of course, just using `auto` with no trailing return is the cleanest, but that requires C++14.

Reply:
Use `value_type` instead of `double` here.